python/recipe.cpp

```diff
@@ -0,0 +1,234 @@
+
+    // Get the python object pyarb::cell_description from the python front end
+    pybind11::object o = impl_->cell_description(gid);
+
+    if (isinstance<arb::cable_cell>(o)) {
+        return arb::util::unique_any(cast<arb::cable_cell>(o));
+    }
+
+    else if (isinstance<arb::lif_cell>(o)) {
+        return arb::util::unique_any(cast<arb::lif_cell>(o));
+    }
+
+    else if (isinstance<arb::spike_source_cell>(o)) {
+        return arb::util::unique_any(cast<arb::spike_source_cell>(o));
+    }
+
+    else if (isinstance<arb::benchmark_cell>(o)) {
+        return arb::util::unique_any(cast<arb::benchmark_cell>(o));
+    }
+
+    throw pyarb_error(
+                        "recipe.cell_description returned \""
+                        + std::string(pybind11::str(o))
+                        + "\" which does not describe a known Arbor cell type");
+}
+
+// The py::recipe::global_properties returns a pybind11::object, that is
+// unwrapped and copied into a arb::util::any.
+arb::util::any py_recipe_shim::get_global_properties(arb::cell_kind kind) const {
+    using pybind11::isinstance;
+    using pybind11::cast;
+
+    // Aquire the GIL because it must be held when calling isinstance and cast.
+    auto guard = pybind11::gil_scoped_acquire();
+
+    // Get the python object pyarb::global_properties from the python front end
+    pybind11::object o = impl_->global_properties(kind);
+
+    if (kind == arb::cell_kind::cable) {
+        return arb::util::any(cast<arb::cable_cell_global_properties>(o));
+    }
+
+    else return arb::util::any{};
+
+    throw pyarb_error(
+                        "recipe.global_properties returned \""
+                        + std::string(pybind11::str(o))
+                        + "\" which does not describe a known Arbor global property description");
+
+}
+
+std::vector<arb::event_generator> py_recipe_shim::event_generators(arb::cell_gid_type gid) const {
+    using namespace std::string_literals;
+    using pybind11::isinstance;
+    using pybind11::cast;
+
+    // Aquire the GIL because it must be held when calling isinstance and cast.
+    auto guard = pybind11::gil_scoped_acquire();
+
+    // Get the python list of pyarb::event_generator_shim from the python front end.
+    auto pygens = impl_->event_generators(gid);
+
+    std::vector<arb::event_generator> gens;
+    gens.reserve(pygens.size());
+
+    for (auto& g: pygens) {
+        // check that a valid Python event_generator was passed.
+        if (!isinstance<pyarb::event_generator_shim>(g)) {
+            std::stringstream s;
+            s << "recipe supplied an invalid event generator for gid "
+            << gid << ": " << pybind11::str(g);
+            throw pyarb_error(s.str());
+        }
+        // get a reference to the python event_generator
+        auto& p = cast<const pyarb::event_generator_shim&>(g);
+
+        // convert the event_generator to an arb::event_generator
+        gens.push_back(
```

Comment:
The indentation here doesn't shorten the lines.

Reply:
One line now.